1.13.1 rc1 cherrypicks

vieux opened the pull request:
Add zsh completion for 'docker {container,image,network,system} prune --filter #30317 
Add zsh completion for 'docker swarm --availability' #30282
Add bash completion for `dockerd --init` and `docker run|create --init` #30266
Add bash completion for `build --squash` #30200
Honour $DOCKER_HIDE_LEGACY_COMMANDS in bash completion #30132
Fix treatment of DOCKER_HIDE_LEGACY_COMMANDS in bash completion #30163
Add bash completion for `docker swarm unlock|unlock-key` #30399
Add missing options to bash completion for `docker swarm init|update` #30398
Add bash completion for `docker service ps --filter node` #30375
Fix bash completion for `docker service update` #30373
Add zsh completion for 'docker plugin install --alias' #30268
Add bash completion for `plugin install --disable-content-trust` #30125
Add zsh completion for 'dockerd --init' and 'docker {create,run} --init' #30318
Ensure proper value is used when computing reclaimable space #30378
Add bash completion and docs for `docker run|create --init-path` #30283
distribution: Add text/html and application/json as image mediatypes #30262
Remove timeout on fifos opening #30140
Ignore certificate expiry error for top-level inspect #29246
Fix pluginv1 Windows volumes #30150
Only show global warnings once #28939
do not create init-dir if not needed #29846
Fix deadlock on v1 plugin with activate error #30408
Fix failure in `docker ps --format` when `.Label` has args #30291
Fix bash completion for completing nodes #30376
plugins: add example docs for a rexray plugin #30294
[project] bump go to go 1.7.4 #29142
[ppc64le] use official go 1.7.4 binaries #29256
Windows: Bump to latest version of git #29159
Windows: Add service dependency ConDrv #30111
Fix incorrect `Scope` in `network ls/inspect` with duplicate network names #30250
Add "src" alias for `--secret` #30407
Fixed secret creation usage during service create in docs #30370
Remove attachable network on swarm leave #30157

glensc committed:
do not require custom build of tini

https://github.com/krallin/tini/issues/55#issuecomment-260507562
https://github.com/krallin/tini/issues/55#issuecomment-260538243
https://github.com/docker/docker/pull/28037

Signed-off-by: Elan Ruusamäe <glen@delfi.ee>
(cherry picked from commit d7df7315971653e11bab124800a5938e7e5c6316)
Signed-off-by: Victor Vieux <vieux@docker.com>

Commit:
Only show global warnings once

Upon each container create I'm seeing these warning **every** time in the
daemon output:
```
WARN[0002] Your kernel does not support swap memory limit
WARN[0002] Your kernel does not support cgroup rt period
WARN[0002] Your kernel does not support cgroup rt runtime
```
Showing them for each container.create() fills up the logs and encourages
people to ignore the output being generated - which means its less likely
they'll see real issues when they happen.  In short, I don't think we
need to show these warnings more than once, so let's only show these
warnings at daemon start-up time.

Signed-off-by: Doug Davis <dug@us.ibm.com>
(cherry picked from commit ff42a2eb41a86217a440a1c4b2afd3c4cd1d48ac)
Signed-off-by: Victor Vieux <vieux@docker.com>

tophj-ibm committed:
[project] bump go to go 1.7.4

Signed-off-by: Christopher Jones <tophj@linux.vnet.ibm.com>
(cherry picked from commit 20e243780ef3b6a7d1debca6279e06b003895121)
Signed-off-by: Victor Vieux <vieux@docker.com>

Commit:
Windows: Bump to latest version of git

Signed-off-by: John Howard <jhoward@microsoft.com>
(cherry picked from commit 8a45eea2ddac40c907feb38859eb056a4d0024cc)
Signed-off-by: Victor Vieux <vieux@docker.com>

tophj-ibm committed:
[ppc64le] use official go 1.7.4 binaries

go 1.7.4 introduced official go ppc64le binaries, so use those
instead of using older versions of go and building from src.

Signed-off-by: Christopher Jones <tophj@linux.vnet.ibm.com>
(cherry picked from commit 62ab6809bd0a654dde9c076f5a197253c66877e8)
Signed-off-by: Victor Vieux <vieux@docker.com>

thaJeztah committed:
Ignore certificate expiry error for top-level inspect

The top-level `docker inspect` command could return
an error if the nodes Swarm certificates were expired.

In situations where the user did not explicitly
ask for an object-type (`--type=foo`), we should
ignore these errors, and consider them equal to
"node is not a swarm manager".

This change makes `docker inspect` ignore these
errors if no type was specified.

As a further optimization, the "swarm status"
result is now stored in a variable, so that
other swarm-specific API calls can be skipped.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit 66cf4cc7f7942b049e0c99c21b84ab07aa80c777)
Signed-off-by: Victor Vieux <vieux@docker.com>

sdurrheimer committed:
Add zsh completion for 'docker plugin install --alias'

Signed-off-by: Steve Durrheimer <s.durrheimer@gmail.com>
(cherry picked from commit 455b7bcbabd4eb7e440448b5834f626cb08c69e0)
Signed-off-by: Victor Vieux <vieux@docker.com>

thaJeztah committed:
do not create init-dir if not needed

commit 56f77d5ade945b3b8816a6c8acb328b7c6dce9a7
added support for cpu-rt-period and cpu-rt-runtime,
but always initialized the cgroup path, even if not
used.

As a result, containers failed to start on a
read-only filesystem.

This patch only creates the cgroup path if
one of these options is set.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit f285d5b3e8eeea7d85e143d845e85a4d4e4c936a)
Signed-off-by: Victor Vieux <vieux@docker.com>

darstahl committed:
Windows: Add service dependency ConDrv

Signed-off-by: Darren Stahl <darst@microsoft.com>
(cherry picked from commit b2a7f6abb379e96eb611c27a5bec10c3be854a75)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Add bash completion for `plugin install --disable-content-trust`

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit 725a864a0496a6c10f56c30f8a45234230951343)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Honour $DOCKER_HIDE_LEGACY_COMMANDS in bash completion

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit aa006f128065fa8b1f017d1fdfdfebdbd08e31c9)
Signed-off-by: Victor Vieux <vieux@docker.com>

mlaventure committed:
Remove timeout on fifos opening

Instead of a timeout the context is cancelled on error to ensure
proper cleanup of the associated fifos' goroutines.

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>
(cherry picked from commit c178700a0420118fe7f632af4da1bc43abd3a9bf)
Signed-off-by: Victor Vieux <vieux@docker.com>

anusha-ragunathan committed:
Fix pluginv1 Windows volumes

c54b717 caused a regression for pluginv1 on Windows, where extraneous
backslashes were added to BasePath of the plugin. For pluginv1 on windows,
BasePath() should return an empty string, since the plugin is fully aware
of the mount path. Also, unlike Linux where all paths are relative to "/",
Windows paths are dependent on system drives and mapped drives.

Fixes #30148

Signed-off-by: Anusha Ragunathan <anusha.ragunathan@docker.com>
(cherry picked from commit 0ef21eb0e30d2ea036730a7c5502f888c6b763d1)
Signed-off-by: Victor Vieux <vieux@docker.com>

aboch committed:
Remove attachable network on swarm leave

- When the node leaves the cluster, if any user run
  container(s) is connected to the swarm network,
  then daemon needs to detach the container(s) and
  remove the network.

Signed-off-by: Alessandro Boch <aboch@docker.com>
(cherry picked from commit 3cedca5d532958ffc007d9b62cc871d3d113f054)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Fix treatment of DOCKER_HIDE_LEGACY_COMMANDS in bash completion

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit 12952f537cf1851bfe84706a4e0eb283dc970379)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Add bash completion for `build --squash`

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit e91dbba7ff62a92bd395bc419f522fa66618ae7b)
Signed-off-by: Victor Vieux <vieux@docker.com>

yongtang committed:
Fix incorrect `Scope` in `network ls/inspect` with duplicate network names

This fix tries to address the issue raised in 30242 where the `Scope`
field always changed to `swarm` in the ouput of `docker network ls/inspect`
when duplicate networks name exist.

The reason for the issue was that `buildNetworkResource()` use network name
(which may not be unique) to check for the scope.

This fix fixes the issue by always use network ID in `buildNetworkResource()`.

A test has been added. The test fails before the fix and passes after the fix.

This fix fixes 30242.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>
(cherry picked from commit 05a831a775be5e8d752deaef620e629deb15cb89)
Signed-off-by: Victor Vieux <vieux@docker.com>

Commit:
distribution: Add text/html and application/json as image mediatypes

As noted by #30083, the new strict checking of mediatypes misses some
cases where earlier bugs caused nonstandard mediatypes to be stored in
manifests. Two of the known cases are text/html and application/json,
which were returned by certain registries and stored by earlier versions
of Docker. Add special cases for text/html and application/json.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>
(cherry picked from commit a215e15cb1fbecc3b22d4f90e15638728ac7ac78)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Add bash completion for `dockerd --init` and `docker run|create --init`

Fixes #30263

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit 20762617572f2e02b8e397844343083506144e86)
Signed-off-by: Victor Vieux <vieux@docker.com>

sdurrheimer committed:
Add zsh completion for 'docker swarm --availability'

Signed-off-by: Steve Durrheimer <s.durrheimer@gmail.com>
(cherry picked from commit ee2ce82c61c11b1c14f9febf811fbbc19d69b750)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Add bash completion for `run|create --init-path`

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit 182048f18592259633d21ac55bd5b73903de50b9)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Add docs for `run|create --init|--init-path`

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit d049ef2b0db4aebfba4f85f5d8294e9884a4b7e2)
Signed-off-by: Victor Vieux <vieux@docker.com>

yongtang committed:
Fix failure in `docker ps --format` when `.Label` has args

This fix tries to fix the issue in 30279 where  `docker ps --format`
fails if `.Label` has args. For example:
```
docker ps --format '{{.ID}}\t{{.Names}}\t{{.Label "some.label"}}'
```

The reason for the failure is that during the preprocessing phase
to detect the existance of `.Size`, the `listOptionsProcessor`
does not has a method of `Label(name string) string`.

This results in the failure of
```
template: :1:24: executing "" at <.Label>: Label is not a method but has arguments
```

This fix fixes the issue by adding needed method of `Label(name string) string`.

This fix fixes 30279.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>
(cherry picked from commit 2cd4ba1e56222a938b2edaa7ad7981c9a2bc248f)
Signed-off-by: Victor Vieux <vieux@docker.com>

Commit:
plugins: add example docs for a rexray plugin

Signed-off-by: Tibor Vass <tibor@docker.com>
(cherry picked from commit 5f713cecc5f65ea61db5b5362f26f96de9427eb0)
Signed-off-by: Victor Vieux <vieux@docker.com>

sdurrheimer committed:
Add zsh completion for 'docker {container,image,network,system} prune --filter'

Signed-off-by: Steve Durrheimer <s.durrheimer@gmail.com>
(cherry picked from commit 784ebccffaa8329aee9e88877e6039b5002ad8e3)
Signed-off-by: Victor Vieux <vieux@docker.com>

sdurrheimer committed:
Add zsh completion for 'dockerd --init' and 'docker {create,run} --init'

Signed-off-by: Steve Durrheimer <s.durrheimer@gmail.com>
(cherry picked from commit cdf57782fd6323549e37aaccc3eeaef1b4aa91f3)
Signed-off-by: Victor Vieux <vieux@docker.com>

mikesir87 committed:
Fixed secret creation usage during service create in docs

Using source, the following error is made:
invalid field in secret request: src

Signed-off-by: Michael Irwin <mikesir87@gmail.com>
(cherry picked from commit 9250117d58443a9db052b69457a135b88f97c322)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Fix bash completion for `docker service update`

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit 01ac33ce165835a3ccde7d151d7d2905c8391461)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Add bash completion for `docker service ps --filter node`

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit a39a2bb6c29405148df3e45216f46f54263cd5ea)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Fix bash completion for completing nodes

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit 550947e437fab42a0b661bad5cb2645f61861aa1)
Signed-off-by: Victor Vieux <vieux@docker.com>

thaJeztah commented:
ping @albers @sdurrheimer if there's changes missing 😄

mlaventure committed:
Ensure proper value is used when computing reclaimable space

When Size was reverted to be equal to VirtualSize, the df command
formatter was not correctly updated to account for the change.

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>
(cherry picked from commit dcc1b4baf6d0eb068010d071bc44daf72b12cd4a)
Signed-off-by: Victor Vieux <vieux@docker.com>

sdurrheimer commented:
#30268

And I think I have one or two more things to add.

albers committed:
Add missing options to bash completion for `docker swarm init|update`

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit f93da79dc18a989a727978350cca8e941e9daeca)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers committed:
Add bash completion for `docker swarm unlock|unlock-key`

Signed-off-by: Harald Albers <github@albersweb.de>
(cherry picked from commit 07bd5152b3731c108e2dba57b1f51e6609b55ba3)
Signed-off-by: Victor Vieux <vieux@docker.com>

albers commented:
#30375 was just merged.
#30125 is about a feature in 1.13.0, so it should go into 1.13.1. (I see missing support for a feature in completion as a bug.)

There are some pending PRs, so stay tuned: #30399, #30398, #30376, #30373, #30371, #30283. 
@thaJeztah maybe you can have a look at these?

thaJeztah committed:
Add "src" alias for `--secret`

This patch adds a "src" alias for `--secret`
to be consistent with `--mount`.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
(cherry picked from commit 950658bbb63480de254a166bdeaef1fc8b8fe644)
Signed-off-by: Victor Vieux <vieux@docker.com>

cpuguy83 committed:
Fix deadlock on v1 plugin with activate error

When a plugin has an activation error, it was not being checked in the
`waitActive` loop. This means it will just wait forever for a manifest
to be populated even though it may never come.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>
(cherry picked from commit f2d384fca6fa08da13fdc01c7991e8e35b081198)
Signed-off-by: Victor Vieux <vieux@docker.com>

vieux commented:
@albers #30283 is not for 1.13.1

I added all the other that are merged

thaJeztah commented on contrib/completion/zsh/_docker:
```diff
@@ -541,6 +561,7 @@ __docker_container_subcommand() {
         "($help)*--group=[Set one or more supplementary user groups for the container]:group:_groups"
         "($help -h --hostname)"{-h=,--hostname=}"[Container host name]:hostname:_hosts"
         "($help -i --interactive)"{-i,--interactive}"[Keep stdin open even if not attached]"
+        "($help)--init[Run an init inside the container that forwards signals and reaps processes]" \
```
This trailing backslash should be removed; I think there's a fix in another PR, but that also has other changes, so we should make sure that either that one, or a fix for just this trailing backslash makes it into 1.13.1

vieux replied:
@thaJeztah fixed in 328775ea4f8446394d45a333652d011f5006a360

vieux committed:
removing trailing backslash in zsh completion

Signed-off-by: Victor Vieux <vieux@docker.com>

unclejack commented:
https://github.com/docker/docker/pull/30489 will also need to be cherry picked after the merge.

tiborvass commented:
LGTM

cpuguy83 reviewed (approved):
LGTM

vieux commented:
@unclejack yes I'll take care of it after

vieux commented:
unrelated windows test error.
